### ISSUE TITLE
Skip memoization on one-off (no watch) builds

### ISSUE THREAD

**Issue by jchavarri** [ocaml/dune#9732]
We are trying to identify sources of slowdown in Ahrefs "noop" builds (builds where all artifacts were already generated and only dune "internal" work is done), as they are taking >20s in some cases at the moment on a 256 cores machine.

Initial profiling shows there is quite intense memory pressure, for example, here's a `xctrace` result on macOS showing ~28% of the time is spent on `caml_call_gc`:

![image](https://github.com/ocaml/dune/assets/220424/5cac57e7-392f-4090-b95f-b9ca2404a3b8)

Or here, a memtrace profile on Linux showing that more than 20GB need to be allocated during the noop build, with more than 5GB being live at the end of the build

![image](https://github.com/ocaml/dune/assets/220424/91b1aafe-ce64-47c3-add0-b38b1277a8cd)

If we filter by functions including `Memo` in their name, memtrace shows the memory allocated during the whole execution is ~5GB, or around 25% of the total memory allocated:

<img width="1631" alt="image" src="https://github.com/ocaml/dune/assets/220424/efeec6bc-c436-4059-924e-30d1210dd8a5">

Here is the memtrace file (renamed to add `.txt` so GH allows uploading):

[dune-build-fe-all.ctf.txt](https://github.com/ocaml/dune/files/13937488/dune-build-fe-all.ctf.txt)

Afaiu, the benefits of `Memo` are mostly obtained on watch mode, where Dune might revisit multiple times the same rules. While I suspect in "one-off" builds these benefits have less impact.

I wonder if there could be a way to have Dune bypass memoization on one-off builds? When there's a computation, just proceed with it without further allocations or dependency tracking? If there's a simple way to patch Dune code to test it, I would happily give it a try on our codebase and report feedback. cc @snowleopard 

Thanks!

**Comment by emillon** [comment]
> Afaiu, the benefits of `Memo` are mostly obtained on watch mode, where Dune might revisit multiple times the same rules. While I suspect in "one-off" builds these benefits have less impact.

I don't think that's accurate. Memo is central to the watch mode, but it's also a way to cache all the "pure" computations made by Dune ("pure" in the sense of not executing commands). For example, getting the compilation flags that apply to a given module. This is an old document at that point, but [this blog post](https://dune.build/blog/new-computation-model/) describe the genesis of `Memo`.

Most of Dune's engine assumes that Memo is there to deduplicate work, so I don't think it would work to have a mode where memoized fibers are run directly.

**Comment by jchavarri** [comment]
@emillon thanks for the feedback.

> For example, getting the compilation flags that apply to a given module.

Could you explain more about how `Memo` helps optimizing this case? Afaiu, the rule and compilation flags for a given module will only be generated just once in a "one-off" build, as there is no way to modify files in that mode. Or would compilation flags be calculated every time the module appears as a dependency of other modules?

> Most of Dune's engine assumes that Memo is there to deduplicate work

I guess this is the assumption I am questioning. I will turn the question around: is there a way to see the total number of operations performed in a build and the percentage of them are being memoized? If there was a way to see this number, and the percentage is too low (let's say < 5%), would it make sense to have this new mode?

**Comment by snowleopard** [comment]
Yeah, agreed with Etienne. In addition to memoization, Memo also detects cycles and has quite a lot of logic for reporting errors.

We are doing some performance improvements in Memo internally, so maybe some of the hot spots you're seeing will soon get improved.

One could imagine turning some features of Memo off for non-incremental builds but I would recommend to postpone that until our general performance improvements land (in 2-3 months).

**Comment by snowleopard** [comment]
Btw, it's not obvious to me that you're observing Memo overhead. More likely, you're hitting the overheads of `Action_builder` and `Dep` modules, which we are improving too.

**Comment by Khady** [comment]
Is there an issue/discussion where those performance improvements are being discussed? We could share some perf/magic-trace data if that helps.

**Comment by davesnx** [comment]
> Btw, it's not obvious to me that you're observing Memo overhead. More likely, you're hitting the overheads of `Action_builder` and `Dep` modules, which we are improving too.

What's the best way to detect which parts are we looking at here? or what's the best way to discard memo from being an overhead?

**Comment by jchavarri** [comment]
> One could imagine turning some features of Memo off for non-incremental builds but I would recommend to postpone that until our general performance improvements land (in 2-3 months).

Thanks for the information, excited about this work! As @Khady mentioned, it'd be great to know more about it.

> Btw, it's not obvious to me that you're observing Memo overhead. More likely, you're hitting the overheads of `Action_builder` and `Dep` modules, which we are improving too.

I might have been misled by the report I shared above, which shows `Memo` functions like `add_path_to` allocating 793MB.

Leaving memory pressure aside, I can see chunks of work originating from `Memo` functions taking > 8% of the time in the build in total, while `Dep` or `Action_builder` seem to take less time. But maybe I'm missing parts of the stacks in the reports I'm taking.

All image captures generated from speedscope, using same build, highlighting the pieces of work that include functions with text in the title:

`Memo.`:
<img width="927" alt="image" src="https://github.com/ocaml/dune/assets/220424/dfa6c2bf-444f-46a1-92c0-f04121d05fd9">

`Dep.`:

<img width="927" alt="image" src="https://github.com/ocaml/dune/assets/220424/ec4bda08-d2f2-4f52-8a1e-e659fd4d27f4">

`Action_builder`:

<img width="928" alt="image" src="https://github.com/ocaml/dune/assets/220424/00fa58f4-ef6e-4030-b11f-032a7ccc6b36">

**Comment by snowleopard** [comment]
> Is there an issue/discussion where those performance improvements are being discussed?

@Khady Sorry, I don't think there is anything public. To make Dune scale to our monorepo with 50 millions lines of OCaml (and 3.5 million rules in total), we had to do a lot of changes in the build engine. Many of them grew out of issues specific to our monorepo, rules and the CI, and it would have been awkward to discuss these issues externally. We've been gradually upstreaming all of these changes, but there is quite a bit of backlog that we need to go through. To add some colour, just today we've released a change internally that speeds up some of our builds by 3.5x! We believe that this has finally brought us on par with Jenga (the build system from which we're switching to Dune).

@jchavarri I didn't mean to say your analysis was wrong, merely that it was not obvious to me from the screenshots above that Memo was the culprit. Speaking of `add_path_to` specifically:

* The code around this function changed somewhat internally, so maybe some of the overheads are gone.
* This function is responsible for cycle detection, so you won't really be able to turn it off in one-shot builds.
